clients/hadoopfs/src/main/java/io/lakefs/FileSystemTracer.java

```diff
@@ -0,0 +1,489 @@
+        } catch (IOException e) {
+            s3aException = e;
+            LOG.error("[mkdirs] Can't mkdir {} to {} with S3AFileSystem, exception {}", f, e.getMessage());
+        }
+
+        if (useLakeFSFileSystemResults && lakeFSException != null) {
+            LOG.trace("[mkdirs] exception by lakeFSFileSystem");
+            throw lakeFSException;
+        }
+        if (!useLakeFSFileSystemResults && s3aException != null) {
+            LOG.trace("[mkdirs] exception by S3AFileSystem");
+            throw s3aException;
+        }
+
+        LOG.trace("{}[mkdirs] lakefs: {}, s3a: {}", RESULTS_COMPARISON, lakefsRes, s3aRes);
+        if (useLakeFSFileSystemResults) {
+            return lakefsRes;
+        }
+        return s3aRes;
+    }
+
+    @Override
+    public FileStatus getFileStatus(Path f) throws IOException {
+        LOG.trace("getFileStatus(f {})", f);
+
+        FileStatus lakefsRes = null;
+        FileStatus s3aRes = null;
+        IOException lakeFSException = null;
+        IOException s3aException = null;
+        Path s3aPath = translateLakeFSPathToS3APath(f);
+        try {
+            lakefsRes = lfsFileSystem.getFileStatus(f);
+        } catch (IOException e) {
+            lakeFSException = e;
+            LOG.error("[getFileStatus] Can't get {} file status with lakeFSFileSystem, exception {}", f, e.getMessage());
+        }
+        try {
+            s3aRes = s3AFileSystem.getFileStatus(s3aPath);
+        } catch (IOException e) {
+            s3aException = e;
+            LOG.error("[getFileStatus] Can't get {} file status with S3AFileSystem, exception {}", s3aPath, e.getMessage());
+        }
+
+        if (useLakeFSFileSystemResults && lakeFSException != null) {
+            LOG.trace("[getFileStatus] exception by lakeFSFileSystem");
+            throw lakeFSException;
+        }
+        if (!useLakeFSFileSystemResults && s3aException != null) {
+            LOG.trace("[getFileStatus] exception by S3AFileSystem");
+            throw s3aException;
+        }
+
+        LOG.trace("{}[getFileStatus] lakefs: {}, s3a: {}", RESULTS_COMPARISON, lakefsRes, s3aRes);
+        if (useLakeFSFileSystemResults) {
+            return lakefsRes;
+        }
+        Path lfsPath = translateS3APathToLakeFSPath(s3aPath);
+        s3aRes.setPath(lfsPath);
+        return s3aRes;
+    }
+
+    /**
+     * An output stream encapsulating two output streams, one for each file system the tracer uses.
+     */
+    private class TracerOutputTStream extends OutputStream {
+
+        private FSDataOutputStream lakeFSStream;
+        private FSDataOutputStream s3aStream;
+
+        public TracerOutputTStream(FSDataOutputStream lakeFSStream, FSDataOutputStream s3aStream) throws IOException {
+            this.lakeFSStream = lakeFSStream;
+            this.s3aStream = s3aStream;
+        }
+
+        @Override
+        public void write(int b) throws IOException {
+            lakeFSStream.write(b);
+            s3aStream.write(b);
```

Comment:
For both streams need to check for null before write - same for other operations.
Stream null is a valid option.